src/java/org/apache/cassandra/db/commitlog/DirectIOSegment.java

```diff
@@ -0,0 +1,216 @@
+ */
+public class DirectIOSegment extends CommitLogSegment
+{
+    private final int fsBlockSize;
+    private final int fsBlockQuotientMask;
+    private final int fsBlockRemainderMask;
+
+    // Needed to track number of bytes written to disk in multiple of page size.
+    long lastWritten = 0;
+
+    /**
+     * Constructs a new segment file.
+     */
+    DirectIOSegment(AbstractCommitLogSegmentManager manager, ThrowingFunction<Path, FileChannel, IOException> channelFactory, int fsBlockSize)
+    {
+        super(manager, channelFactory);
+
+        assert Integer.highestOneBit(fsBlockSize) == fsBlockSize : "fsBlockSize must be a power of 2";
+
+        // mark the initial sync marker as uninitialised
+        int firstSync = buffer.position();
+        buffer.putInt(firstSync + 0, 0);
+        buffer.putInt(firstSync + 4, 0);
+
+        this.fsBlockSize = fsBlockSize;
+        this.fsBlockRemainderMask = fsBlockSize - 1;
+        this.fsBlockQuotientMask = ~fsBlockRemainderMask;
+    }
+
+    @Override
+    void writeLogHeader()
+    {
+        super.writeLogHeader();
+        // Testing shows writing initial bytes takes some time for Direct I/O. During peak load,
+        // it is better to make "COMMIT-LOG-ALLOCATOR" thread to write these few bytes of each
+        // file and this helps syncer thread to speedup the flush activity.
+        flush(0, lastSyncedOffset);
+    }
+
+    @Override
+    void write(int startMarker, int nextMarker)
+    {
+        // if there's room in the discard section to write an empty header,
+        // zero out the next sync marker so replayer can cleanly exit
+        if (nextMarker <= buffer.capacity() - SYNC_MARKER_SIZE)
+        {
+            buffer.putInt(nextMarker, 0);
+            buffer.putInt(nextMarker + 4, 0);
+        }
+
+        // write previous sync marker to point to next sync marker
+        // we don't chain the crcs here to ensure this method is idempotent if it fails
+        writeSyncMarker(id, buffer, startMarker, startMarker, nextMarker);
+    }
+
+    @Override
+    protected void flush(int startMarker, int nextMarker)
+    {
+        try
+        {
+            // lastSyncedOffset is synced to disk. Align lastSyncedOffset to start of its block
+            // and nextMarker to end of its block to avoid write errors.
+            int flushPosition = lastSyncedOffset;
+            ByteBuffer duplicate = buffer.duplicate();
+
+            // Aligned file position if not aligned to start of a block.
+            if ((flushPosition & fsBlockRemainderMask) != 0)
+            {
+                flushPosition = flushPosition & fsBlockQuotientMask;
+                channel.position(flushPosition);
+            }
+            duplicate.position(flushPosition);
+
+            int flushLimit = nextMarker;
+
+            // Align last byte to end of block.
+            if ((flushLimit & fsBlockRemainderMask) != 0)
+                flushLimit = (flushLimit + fsBlockSize) & fsBlockQuotientMask;
```

Comment:
Nit: Remove the `if` and do this as 
`flushLimit = (flushLimit + fsBlockSize - 1) & fsBlockQuotientMask;`
or 
`flushLimit = (flushLimit | (fsBlockSize - 1)) + 1`.

`fsBlockRemainderMask` can be removed; `fsBlockQuotientMask` is `-fsBlockSize`. FWIW calculating these is faster and safer than fetching them, though possibly not as readable. At some later time it would be best to extend `PageAware` to offer methods that apply to arbitrary page sizes.

Reply:
I'll use the first one; I'm not sure if the second one would work without the "if" - it looks like it would unnecessarily move forward the flush limit if it is already aligned